c++/src/kj/compat/http.c++

```diff
@@ -5204,65 +5960,77 @@ private:
+          }
+        } else {
+          connectionHeaders[HttpHeaders::BuiltinIndices::TRANSFER_ENCODING] = "chunked";
+        }
 
-    httpOutput.writeHeaders(headers.serializeResponse(
-        statusCode, statusText, connectionHeadersArray));
+        // For HEAD requests, if the application specified a Content-Length or Transfer-Encoding
+        // header, use that instead of whatever we decided above.
+        kj::ArrayPtr<kj::StringPtr> connectionHeadersArray = connectionHeaders;
+        if (m == HttpMethod::HEAD) {
+          if (headers.get(HttpHeaderId::CONTENT_LENGTH) != nullptr ||
+              headers.get(HttpHeaderId::TRANSFER_ENCODING) != nullptr) {
+            connectionHeadersArray = connectionHeadersArray
+                .slice(0, HttpHeaders::HEAD_RESPONSE_CONNECTION_HEADERS_COUNT);
+          }
+        }
 
-    kj::Own<kj::AsyncOutputStream> bodyStream;
-    if (method == HttpMethod::HEAD) {
-      // Ignore entity-body.
-      httpOutput.finishBody();
-      return heap<HttpDiscardingEntityWriter>();
-    } else if (statusCode == 204 || statusCode == 205 || statusCode == 304) {
-      // No entity-body.
-      httpOutput.finishBody();
-      return heap<HttpNullEntityWriter>();
-    } else KJ_IF_MAYBE(s, expectedBodySize) {
-      return heap<HttpFixedLengthEntityWriter>(httpOutput, *s);
-    } else {
-      return heap<HttpChunkedEntityWriter>(httpOutput);
+        httpOutput.writeHeaders(headers.serializeResponse(
+            statusCode, statusText, connectionHeadersArray));
+
+        kj::Own<kj::AsyncOutputStream> bodyStream;
+        if (m == HttpMethod::HEAD) {
+          // Ignore entity-body.
+          httpOutput.finishBody();
+          return heap<HttpDiscardingEntityWriter>();
+        } else if (statusCode == 204 || statusCode == 205 || statusCode == 304) {
+          // No entity-body.
+          httpOutput.finishBody();
+          return heap<HttpNullEntityWriter>();
+        } else KJ_IF_MAYBE(s, expectedBodySize) {
+          return heap<HttpFixedLengthEntityWriter>(httpOutput, *s);
+        } else {
+          return heap<HttpChunkedEntityWriter>(httpOutput);
+        }
+      }
+      KJ_CASE_ONEOF(m, HttpConnectMethod) {
+        httpOutput.writeHeaders(headers.serializeResponse(
+            statusCode, statusText, connectionHeaders));
+        return heap<HttpPassthroughWriter>(httpOutput);
+      }
     }
+    KJ_UNREACHABLE;
   }
 
   kj::Own<WebSocket> acceptWebSocket(const HttpHeaders& headers) override {
     auto& requestHeaders = httpInput.getHeaders();
     KJ_REQUIRE(requestHeaders.isWebSocket(),
         "can't call acceptWebSocket() if the request headers didn't have Upgrade: WebSocket");
 
-    auto method = KJ_REQUIRE_NONNULL(currentMethod, "already called send()");
+    auto method = KJ_REQUIRE_NONNULL(
+          KJ_REQUIRE_NONNULL(currentMethod, "already called send()")
+              .tryGet<HttpMethod>(), "WebSockets are not supported when using CONNECT");
```

Comment:
The very next check below is verifying the method was GET. Maybe that check should be modified instead of adding an extra check here?